File: lib_local/gallery/src/main/java/com/yanzhenjie/durban/util/FastBitmapDrawable.java
package com.yanzhenjie.durban.util;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.drawable.Drawable;

import androidx.annotation.NonNull;

/**
 * 高效 Bitmap 绘制 Drawable
 * 作用：裁剪界面中专门用来显示图片的高性能组件
 * 比系统 BitmapDrawable 更轻、更快、无内存抖动
 */
public class FastBitmapDrawable extends Drawable {
    private int mAlpha;
    private int mWidth, mHeight;
    private Bitmap mBitmap;
    // 图片抗锯齿画笔
    private final Paint mPaint = new Paint(Paint.FILTER_BITMAP_FLAG);

    public FastBitmapDrawable(Bitmap b) {
        mAlpha = 255;
        setBitmap(b);
    }

    @Override
    public void draw(@NonNull Canvas canvas) {
        if (mBitmap != null && !mBitmap.isRecycled()) {
            canvas.drawBitmap(mBitmap, null, getBounds(), mPaint);
        }
    }

    @Override
    public void setColorFilter(ColorFilter cf) {
        mPaint.setColorFilter(cf);
    }

    @Override
    public int getOpacity() {
        return PixelFormat.TRANSLUCENT;
    }

    @Override
    public void setAlpha(int alpha) {
        mAlpha = alpha;
        mPaint.setAlpha(alpha);
    }

    @Override
    public int getIntrinsicWidth() {
        return mWidth;
    }

    @Override
    public int getIntrinsicHeight() {
        return mHeight;
    }

    @Override
    public int getMinimumWidth() {
        return mWidth;
    }

    @Override
    public int getMinimumHeight() {
        return mHeight;
    }

    public void setFilterBitmap(boolean filterBitmap) {
        mPaint.setFilterBitmap(filterBitmap);
    }

    public int getAlpha() {
        return mAlpha;
    }

    public void setBitmap(Bitmap b) {
        mBitmap = b;
        if (b != null) {
            mWidth = mBitmap.getWidth();
            mHeight = mBitmap.getHeight();
        } else {
            mWidth = mHeight = 0;
        }
    }

    public Bitmap getBitmap() {
        return mBitmap;
    }

}

File: lib_local/gallery/src/main/java/com/yanzhenjie/durban/widget/FastBitmapDrawable.kt
package com.yanzhenjie.durban.widget

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.ColorFilter
import android.graphics.Paint
import android.graphics.PixelFormat
import android.graphics.drawable.Drawable

/**
 * 高效 Bitmap 绘制 Drawable
 * 作用：裁剪界面中专门用来显示图片的高性能组件
 */
class FastBitmapDrawable(b: Bitmap) : Drawable() {
    private var mWidth = 0
    private var mHeight = 0
    private var mAlpha = 255
    // 图片抗锯齿画笔
    private val mPaint = Paint(Paint.FILTER_BITMAP_FLAG)
    private var mBitmap: Bitmap? = null

    init {
        setBitmap(b)
    }

    override fun draw(canvas: Canvas) {
        mBitmap?.takeIf { !it.isRecycled }?.let { bitmap ->
            canvas.drawBitmap(bitmap, null, bounds, mPaint)
        }
    }

    @Deprecated("Deprecated in Java")
    override fun getOpacity(): Int {
        return PixelFormat.TRANSLUCENT
    }

    override fun setAlpha(alpha: Int) {
        mAlpha = alpha
        mPaint.setAlpha(alpha)
    }

    override fun setColorFilter(cf: ColorFilter?) {
        mPaint.setColorFilter(cf)
    }

    override fun setFilterBitmap(filterBitmap: Boolean) {
        mPaint.isFilterBitmap = filterBitmap
    }

    override fun getIntrinsicWidth(): Int {
        return mWidth
    }

    override fun getIntrinsicHeight(): Int {
        return mHeight
    }

    override fun getMinimumWidth(): Int {
        return mWidth
    }

    override fun getMinimumHeight(): Int {
        return mHeight
    }

    fun setBitmap(b: Bitmap?) {
        mBitmap = b
        mWidth = mBitmap?.width ?: 0
        mHeight = mBitmap?.height ?: 0
    }

    fun getBitmap(): Bitmap? {
        return mBitmap
    }

    override fun getAlpha(): Int {
        return mAlpha
    }

}